### ISSUE TITLE
VGGFace2 fine-tune: Poor training accuracy on CPU

### ISSUE THREAD

**Issue by AliAmjad** [timesler/facenet-pytorch#104]
I'm using this code to fine-tune the model but I'm getting very poor training accuracy, could you please point out what could be the issue:

    # define mtcnn
    mtcnn = MTCNN(
        image_size=160, margin=0, min_face_size=20,
        thresholds=[0.6, 0.7, 0.7], factor=0.709, post_process=True,
        device=device
    )

    # load data
    dataset = datasets.ImageFolder(data_dir, transform=transforms.Resize((512, 512)))

    dataset.samples = [
        (p, p.replace(data_dir, data_dir + '_cropped'))
            for p, _ in dataset.samples
    ]

    loader = DataLoader(
        dataset,
        num_workers=workers,
        batch_size=batch_size,
        collate_fn=training.collate_pil
    )

    print('applying mtcnn...')

    # apply mtcnn
    for i, (x, y) in enumerate(loader):
        mtcnn(x, save_path=y)
        print('\rBatch {} of {}'.format(i + 1, len(loader)), end='')
        
    # Remove mtcnn to reduce GPU memory usage
    del mtcnn

    print('')
    print('starting training...')

    print(len(dataset.class_to_idx))

    resnet = InceptionResnetV1(
        classify=False,
        pretrained='vggface2',
        num_classes=len(dataset.class_to_idx)
    ).to(device)

    optimizer = optim.Adam(resnet.parameters(), lr=0.001)
    scheduler = MultiStepLR(optimizer, [5, 10])

    trans = transforms.Compose([
        np.float32,
        transforms.ToTensor(),
        fixed_image_standardization
    ])

    dataset = datasets.ImageFolder(data_dir + '_cropped', transform=trans)
    img_inds = np.arange(len(dataset))
    np.random.shuffle(img_inds)
    train_inds = img_inds[:int(0.8 * len(img_inds))]
    val_inds = img_inds[int(0.8 * len(img_inds)):]

    train_loader = DataLoader(
        dataset,
        num_workers=workers,
        batch_size=batch_size,
        sampler=SubsetRandomSampler(train_inds)
    )

    val_loader = DataLoader(
        dataset,
        num_workers=workers,
        batch_size=batch_size,
        sampler=SubsetRandomSampler(val_inds)
    )

    loss_fn = torch.nn.CrossEntropyLoss()
    metrics = {
        'fps': training.BatchTimer(),
        'acc': training.accuracy
    }

    writer = SummaryWriter()
    writer.iteration, writer.interval = 0, 10

    print('\n\nInitial')
    print('-' * 10)
    resnet.eval()
    training.pass_epoch(
        resnet, loss_fn, val_loader,
        batch_metrics=metrics, show_running=True, device=device,
        writer=writer
    )

    for epoch in range(epochs):
        print('\nEpoch {}/{}'.format(epoch + 1, epochs))
        print('-' * 10)

        resnet.train()
        training.pass_epoch(
            resnet, loss_fn, train_loader, optimizer, scheduler,
            batch_metrics=metrics, show_running=True, device=device,
            writer=writer
        )

        resnet.eval()
        training.pass_epoch(
            resnet, loss_fn, val_loader,
            batch_metrics=metrics, show_running=True, device=device,
            writer=writer
        )

    writer.close()

    # save trained model
    torch.save(resnet.state_dict(), trained_model)

and this is the accuracy I'm getting after 8 epochs:

Initial
----------
Valid |     1/1    | loss:    6.2234 | fps:    1.4201 | acc:    0.0000   

Epoch 1/8
----------
Train |     1/1    | loss:    6.2504 | fps:    0.8825 | acc:    0.0000   
Valid |     1/1    | loss:    6.2631 | fps:    1.5254 | acc:    0.0000   

Epoch 2/8
----------
Train |     1/1    | loss:    6.1771 | fps:    1.3746 | acc:    0.0000   
Valid |     1/1    | loss:    6.2687 | fps:    1.3014 | acc:    0.0000   

Epoch 3/8
----------
Train |     1/1    | loss:    6.1627 | fps:    1.0087 | acc:    0.2500   
Valid |     1/1    | loss:    6.2471 | fps:    1.5138 | acc:    0.0000   

Epoch 4/8
----------
Train |     1/1    | loss:    6.1570 | fps:    1.0297 | acc:    0.2500   
Valid |     1/1    | loss:    6.2371 | fps:    1.8226 | acc:    0.0000   

Epoch 5/8
----------
Train |     1/1    | loss:    6.1445 | fps:    0.8727 | acc:    0.5000   
Valid |     1/1    | loss:    6.2335 | fps:    0.9244 | acc:    0.0000   

Epoch 6/8
----------
Train |     1/1    | loss:    6.1274 | fps:    1.1550 | acc:    0.5000   
Valid |     1/1    | loss:    6.2234 | fps:    1.6978 | acc:    0.0000   

Epoch 7/8
----------
Train |     1/1    | loss:    6.1252 | fps:    1.4416 | acc:    0.2500   
Valid |     1/1    | loss:    6.1999 | fps:    1.7895 | acc:    0.0000   

Epoch 8/8
----------
Train |     1/1    | loss:    6.1179 | fps:    1.4245 | acc:    0.5000   
Valid |     1/1    | loss:    6.1874 | fps:    1.6070 | acc:    0.0000

**Comment by stech-apps** [comment]
One point to share when I ran this code for the first time it gives the accuracy 100% in less than 8 epochs but after the second run above is the accuracy received.

**Comment by AliAmjad** [comment]
One point to share when I ran this code for the first time it gives the accuracy 100% in less than 8 epochs but after the second run above is the accuracy received.

**Comment by timesler** [comment]
@AliAmjad what dataset are you using? The reason your model is not improving is because you are not training it on any data. The output you have copied shows a single batch in your training set.

**Comment by AliAmjad** [comment]
@timesler I'm using the data in the test_images folder and plus my own face images in separate folder, there are 6 faces in total and I can see the mtcnn being performed and they are in "..._cropped" folder.

**Comment by AliAmjad** [comment]
@timesler I may have found the problem, when you set "classify=False" you get poor accuracy but when it's set to True you get good accuracy, I've used your finetune notebook and you can reproduce it there as well.

**Comment by timesler** [comment]
To finetune the model in a meaningful way, you would need many more images (1000s). When `classify=False` the model returns 512 dimensional embeddings and isn't used for classification. You often don't need to fine tune the model to use it on new faces, you can just calculate distances in the embedding space.

**Comment by kartikpodugu** [comment]
@timesler 
Suppose I don't want to finetune the model. 
I want to use existing model only for recognizing new faces. 
I used the model with  `classify=False`, then I will get the embeddings of the face. 
Then what I have to do?
how do I calculate distance in embedding space for the new face? and how do I recognize the face?
Can you please elaborate?
